src/Orleans.Streaming/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -1,453 +1,498 @@
+    public override IEnumerable<QueueId> GetMyQueues()
+    {
+        if (Cancellation.IsCancellationRequested)
         {
-            if (base.Cancellation.IsCancellationRequested) throw new InvalidOperationException("Cannot aquire queues from a terminated balancer.");
-            return this.myQueues.Select(queue => queue.QueueId);
+            throw new InvalidOperationException("Cannot acquire queues from a terminated balancer.");
         }
 
-        private async Task MaintainLeases(object state)
+        return _myQueues.Select(queue => queue.QueueId);
+    }
+
+    private async Task PeriodicallyMaintainLeases()
+    {
+        await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);
+        while (await _leaseMaintenanceTimer.WaitForNextTickAsync())
         {
             try
             {
-                await this.executor.AddNext(this.MaintainLeases);
+                await _executor.AddNext(MaintainLeases);
             }
             catch (Exception ex)
             {
-                this.Logger.LogError(ex, "Maintaining leases failed");
+                Logger.LogError(ex, "Error maintaining leases.");
             }
         }
 
-        private async Task MaintainLeases()
+        async Task MaintainLeases()
         {
-            if (base.Cancellation.IsCancellationRequested) return;
-            var oldQueues = new HashSet<QueueId>(this.myQueues.Select(queue => queue.QueueId));
+            if (Cancellation.IsCancellationRequested) return;
+            var oldQueues = new HashSet<QueueId>(_myQueues.Select(queue => queue.QueueId));
             try
             {
-                bool allLeasesRenewed = await this.RenewLeases();
-                // if we lost some leases during renew after leaseAquisitionTimer stopped, restart it
-                if (!allLeasesRenewed &&
-                    this.leaseAquisitionTimer == null &&
-                    !base.Cancellation.IsCancellationRequested)
+                bool allLeasesRenewed = await RenewLeases();
+
+                // If we lost some leases during renew after leaseAcquisitionTimer stopped, restart it.
+                if (!allLeasesRenewed)
                 {
-                    this.leaseAquisitionTimer = this.timerRegistry.RegisterTimer(null, this.AcquireLeasesToMeetResponsibility, null, TimeSpan.Zero, this.options.LeaseAquisitionPeriod);
+                    // Make the acquisition timer fire immediately.
+                    _leaseAcquisitionTimer.Period = TimeSpan.FromMilliseconds(1);
                 }
             }
             finally
             {
-                await this.NotifyOnChange(oldQueues);
+                await NotifyOnChange(oldQueues);
             }
         }
+    }
 
-        private async Task AcquireLeasesToMeetResponsibility(object state)
+    private async Task PeriodicallyAcquireLeasesToMeetResponsibility()
+    {
+        await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);
```

Comment:
Is this better than just
```suggestion
        await Task.Yield();
```
?

Reply:
For reference: https://github.com/dotnet/runtime/pull/87067/files#r1215280841